aiohttp/web_protocol.py

```diff
@@ -343,40 +342,38 @@ def data_received(self, data: bytes) -> None:
         if self._force_close or self._close:
             return
         # parse http messages
+        messages: Sequence[Tuple[Union[RawRequestMessage, _ErrInfo], StreamReader]]
         if self._payload_parser is None and not self._upgrade:
             assert self._request_parser is not None
             try:
                 messages, upgraded, tail = self._request_parser.feed_data(data)
             except HttpProcessingError as exc:
-                # something happened during parsing
-                self._error_handler = self._loop.create_task(
-                    self.handle_parse_error(
-                        StreamWriter(self, self._loop), 400, exc, exc.message
-                    )
-                )
-                self.close()
+                messages = [
+                    (_ErrInfo(status=400, exc=exc, message=exc.message), EMPTY_PAYLOAD)
+                ]
+                upgraded = False
+                tail = b""
             except Exception as exc:
                 # 500: internal error
-                self._error_handler = self._loop.create_task(
-                    self.handle_parse_error(StreamWriter(self, self._loop), 500, exc)
-                )
-                self.close()
-            else:
-                if messages:
-                    # sometimes the parser returns no messages
-                    for (msg, payload) in messages:
-                        self._request_count += 1
-                        self._messages.append((msg, payload))
-
-                    waiter = self._waiter
-                    if waiter is not None:
-                        if not waiter.done():
-                            # don't set result twice
-                            waiter.set_result(None)
-
-                self._upgrade = upgraded
-                if upgraded and tail:
-                    self._message_tail = tail
+                messages = [(_ErrInfo(status=500, exc=exc), EMPTY_PAYLOAD)]
+                upgraded = False
+                tail = b""
+
+            if messages:
+                # sometimes the parser returns no messages
+                for (msg, payload) in messages:
+                    self._request_count += 1
+                    self._messages.append((msg, payload))
+
+                waiter = self._waiter
+                if waiter is not None:
+                    if not waiter.done():
+                        # don't set result twice
+                        waiter.set_result(None)
```

Comment:
you have a chance to reduce some unnecessary nesting/complexity here:
```suggestion
            for msg, payload in (messages or ()):
                self._request_count += 1
                self._messages.append((msg, payload))

            waiter = self._waiter
            if messages and waiter is not None and not waiter.done():
                # don't set result twice
                waiter.set_result(None)
```